lib/puppet/ssl/ssl_provider.rb

```diff
@@ -0,0 +1,205 @@
+  # @param client_cert [OpenSSL::X509::Certificate] client's cert whose public
+  #   key matches the `private_key`
+  # @param revocation [:chain, :leaf, false] revocation mode
+  # @return [Puppet::SSL::SSLContext] A context to use to create connections
+  # @raise [Puppet::SSL::CertVerifyError] There was an issue with
+  #   one of the certs or CRLs.
+  # @raise [Puppet::SSL::SSLError] There was an issue with the
+  #   `private_key`.
+  # @api private
+  def create_context(cacerts:, crls:, private_key:, client_cert:, revocation: Puppet[:certificate_revocation])
+    raise ArgumentError, "Private key is required" unless private_key
+    raise ArgumentError, "Client cert is required" unless client_cert
+
+    verify_cacerts(cacerts, crls, revocation)
+    store = create_x509_store(cacerts, crls, revocation)
+    client_chain = verify_cert_with_store(store, client_cert)
+
+    unless private_key.is_a?(OpenSSL::PKey::RSA)
+      raise Puppet::SSL::SSLError, _("Unsupported key '%{type}'") % { type: private_key.class.name }
+    end
+
+    unless client_cert.check_private_key(private_key)
+      raise Puppet::SSL::SSLError, _("The certificate for '%{name}' does not match its private key") % { name: subject(client_cert) }
+    end
+
+    Puppet::SSL::SSLContext.new(
+      store: store, trusted_certs: cacerts, crls: crls,
+      private_key: private_key, client_cert: client_cert, client_chain: client_chain
+    )
+  end
+
+  # Verify the `csr` was signed with a private key corresponding to the
+  # `public_key`. This ensures the CSR was signed by someone in possession
+  # of the private key, and that it hasn't been tampered with since.
+  #
+  # @param csr [OpenSSL::X509::Request] certificate signing request
+  # @param public_key [OpenSSL::PKey::RSA] public key
+  # @raise [Puppet::SSL:SSLError] The private_key for the given `public_key` was
+  #   not used to sign the CSR.
+  # @api private
+  def verify_request(csr, public_key)
+    unless csr.verify(public_key)
+      raise Puppet::SSL::SSLError, _("The CSR for host '%{name}' does not match the public key") % { name: subject(csr) }
+    end
+
+    csr
+  end
+
+  private
+
+  def default_flags
+    # checking the signature of the self-signed cert doesn't add any security,
+    # but it's a sanity check to make sure the cert isn't corrupt. This option
+    # is only available in openssl 1.1+
+    if defined?(OpenSSL::X509::V_FLAG_CHECK_SS_SIGNATURE)
+      OpenSSL::X509::V_FLAG_CHECK_SS_SIGNATURE
+    else
+      0
+    end
+  end
+
+  def verify_cacerts(cacerts, crls, revocation)
+    # If revocation mode is :leaf, then we can't perform revocation checking
+    # until we know which issuer is the leaf, so disable revocation checking
+    # while verifying the trusted CA certs.
+    store = create_x509_store(cacerts, crls, revocation == :leaf ? false : revocation)
+
+    cacerts.each do |cert|
+      verify_cert_with_store(store, cert)
+    end
```

Comment:
From today's meeting, there's a chance the ca bundle contains invalid certs that aren't being used for either the client or server chains, such as an expired or revoked intermediate CA. This is especially true given that the agent never updates the CA or CRL bundles automatically. It would be good to only verify CA certs we know are necessary to validate the client's cert, and perhaps warn for invalid certs.

Reply:
Updated, so only `create_context` verifies the cert chain

Reply:
Just to confirm your previous comment above - we are saying will validate _all_ certificates making up the chain for the certificate being validated and just exclude/ignore any extraneous ones that may have been part of the bundle?

Reply:
Correct, we'll verify all certificates in the chain for the client cert. This will be used in the future to verify the client cert we download during the bootstrap process is valid, so that we don't commit an invalid cert to disk and brick the agent.

Note the agent will verify the server's cert and chain, when we later make a connection, but we don't know ahead of time which CAs/CRLs will be needed until later.